src/main/scala/uk/gov/nationalarchives/downloadfiles/FileUtils.scala

```diff
@@ -0,0 +1,73 @@
+package uk.gov.nationalarchives.downloadfiles
+
+import java.net.URLDecoder
+import java.nio.file.Paths
+import java.util.UUID
+
+import com.amazonaws.services.lambda.runtime.events.models.s3.S3EventNotification.S3EventNotificationRecord
+import com.typesafe.config.{Config, ConfigFactory}
+import graphql.codegen.GetOriginalPath.getOriginalPath.{Data, Variables, document}
+import software.amazon.awssdk.services.s3.S3Client
+import software.amazon.awssdk.services.s3.model.GetObjectRequest
+import sttp.client.{Identity, NothingT, SttpBackend}
+import uk.gov.nationalarchives.tdr.{GraphQLClient, GraphQlResponse}
+import uk.gov.nationalarchives.tdr.error.NotAuthorisedError
+import uk.gov.nationalarchives.tdr.keycloak.KeycloakUtils
+
+import scala.concurrent.{ExecutionContext, Future}
+import scala.util.{Failure, Success, Try}
+
+class FileUtils()(implicit val executionContext: ExecutionContext) {
+  val config: Config = ConfigFactory.load
+
+  case class DownloadFilesException(msg: String) extends RuntimeException
+
+  implicit class OptFunction(strOpt: Option[String]) {
+    def failIfPathEmpty(fileId: UUID): Try[String] = if ((strOpt.isDefined && strOpt.get.isEmpty) || strOpt.isEmpty) {
+      Failure(new RuntimeException(s"The original path for fileId $fileId is missing or empty"))
+    } else {
+      Success(strOpt.get)
+    }
+  }
+
+  def getFilePath(keycloakUtils: KeycloakUtils, client: GraphQLClient[Data, Variables], fileId: UUID)(implicit backend: SttpBackend[Identity, Nothing, NothingT]): Future[Try[String]] = {
```

Comment:
A `Future` and a `Try` both have a throwable as the error case, so you should be able to flatten `Future[Try[String]]` to `Future[String]` so that you have fewer layers to deal with in the Lambda code.

One way to do this is to use `Future.fromTry` to create a `Future[Future[String]]` and then flattening it, but there might be a better way.

Reply:
Yeah, you're right. I've refactored it to return Future[String] which has simplified things a bit. Once I have List[Future[String]] in the Lambda class though, I have to turn it back into a Future[List[Try]] because I can't see another way of extracting the Throwable from the future and splitting the Futures into succeeded and failed ones which is what we need.